job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -272,6 +292,50 @@ private void copyLogs() {
             }
         }
     }
+
+    private String getLogPrefix(FlowNode node) {
+        BodyInvocationAction bodyInvocationAction = node.getAction(BodyInvocationAction.class);
+
+        if (bodyInvocationAction != null) {
+            BodyExecutionLabelAction bodyLabelAction = node.getAction(BodyExecutionLabelAction.class);
+
+            if (bodyLabelAction != null) {
+                return bodyLabelAction.getBodyLabel();
+            } else {
+                // This exec body does not have a body label. Stop now.
+                return null;
+            }
+        }
+
+        List<FlowNode> parents = node.getParents();
+        if (parents != null && !parents.isEmpty()) {
+            for (FlowNode parent : parents) {
+                String prefix = getLogPrefix(parent);
+                if (prefix != null) {
+                    return prefix;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    protected static final class LogLinePrefixOutputFilter extends LineTransformationOutputStream {
```

Comment:
Sure .. I made it protected because I initially had something else in mind and thought I'd be writing tests specifically targeting it, which never happened. Thanks for spotting it.